File: sample-conference-videochat/sample-conference-videochat/Components/Calls/BaseViewController/BaseViewController.m
//
//  BaseViewController.m
//  sample-conference-videochat
//
//  Created by Injoit on 21.03.2021.
//  Copyright © 2021 Quickblox. All rights reserved.
//

#import "BaseViewController.h"
#import "CallGradientView.h"

@interface BaseViewController ()<UICollectionViewDelegate, UICollectionViewDataSource>

@property (strong, nonatomic) CallGradientView *containerToolBarView;
@property (strong, nonatomic) CallGradientView *topGradientView;
@property (strong, nonatomic) NSLayoutConstraint *topGradientViewHeightConstraint;
@property (strong, nonatomic) NSLayoutConstraint *containerToolBarTopConstraint;
@property (strong, nonatomic) NSTimer *toolbarHideTimer;

@end

@implementation BaseViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self configureGUI];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    [self setupNavigationBarWillAppear:YES];
    [self showControls:YES];
    [self setupHideToolbarTimerWithTimeInterval:kHideTimeInterval];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    [self setupNavigationBarWillAppear:NO];
    [self invalidateHideToolbarTimer];
}

#pragma mark - These methods can be overridden in child controllers
- (void)configureNavigationBarItems {
    // configure it if necessary.
}

- (void)configureToolBar {
    // configure it if necessary.
}

- (void)setupCollectionView {
    // configure it if necessary.
    UICollectionViewFlowLayout *layout = [[UICollectionViewFlowLayout alloc] init];
    self.collectionView = [[UICollectionView alloc] initWithFrame:CGRectZero collectionViewLayout:layout];
}

- (void)reloadContent {
    // configure it if necessary.
    [self.collectionView reloadData];
}

- (void)configureGUI {
    self.view.backgroundColor = UIColor.blackColor;
    
    // configure it if necessary.
    [self setupCollectionView];
    
    self.collectionView.dataSource = self;
    self.collectionView.delegate = self;
    
    [self.view addSubview:self.collectionView];
    self.collectionView.translatesAutoresizingMaskIntoConstraints = NO;
    CGFloat topBarHeight = self.view.window.windowScene.statusBarManager.statusBarFrame.size.height + self.navigationController.navigationBar.frame.size.height;
    self.collectionView.contentInset = UIEdgeInsetsMake(-topBarHeight, 0.0f, 0.0f, 0.0f);
    self.collectionViewTopConstraint = [self.collectionView.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor];
    self.collectionViewTopConstraint.constant = -topBarHeight;
    self.collectionViewTopConstraint.active = YES;
    [self.collectionView.leftAnchor constraintEqualToAnchor:self.view.leftAnchor].active = YES;
    [self.collectionView.rightAnchor constraintEqualToAnchor:self.view.rightAnchor].active = YES;
    [self.collectionView.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor].active = YES;

    self.containerToolBarView = [[CallGradientView alloc] initWithFrame:CGRectZero];
    [self.view addSubview:self.containerToolBarView];
    self.containerToolBarView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.containerToolBarView.leftAnchor constraintEqualToAnchor:self.view.leftAnchor].active = YES;
    [self.containerToolBarView.rightAnchor constraintEqualToAnchor:self.view.rightAnchor].active = YES;
    [self.containerToolBarView.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor].active = YES;
    self.containerToolBarTopConstraint = [self.containerToolBarView.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor constant: -126.0f];
    self.containerToolBarTopConstraint.active = YES;
    [self.containerToolBarView.heightAnchor constraintEqualToConstant: 126.0f].active = YES;
    self.containerToolBarView.isVertical = YES;

    self.toolbar = [[ToolBar alloc] initWithFrame:CGRectZero];
    [self.containerToolBarView addSubview:self.toolbar];
    self.toolbar.translatesAutoresizingMaskIntoConstraints = NO;
    [self.toolbar.leftAnchor constraintEqualToAnchor:self.containerToolBarView.leftAnchor].active = YES;
    [self.toolbar.rightAnchor constraintEqualToAnchor:self.containerToolBarView.rightAnchor].active = YES;
    [self.toolbar.topAnchor constraintEqualToAnchor:self.containerToolBarView.topAnchor].active = YES;
    [self.toolbar.heightAnchor constraintEqualToConstant: 96.0f].active = YES;

    self.topGradientView = [[CallGradientView alloc] initWithFrame:CGRectZero];
    [self.view addSubview:self.topGradientView];
    self.topGradientView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.topGradientView.leftAnchor constraintEqualToAnchor:self.view.leftAnchor].active = YES;
    [self.topGradientView.rightAnchor constraintEqualToAnchor:self.view.rightAnchor].active = YES;
    [self.topGradientView.topAnchor constraintEqualToAnchor:self.view.topAnchor].active = YES;
    self.topGradientView.isVertical = YES;

    self.topGradientViewHeightConstraint = [self.topGradientView.heightAnchor constraintEqualToConstant: 100.0f];
    if (UIApplication.sharedApplication.windows.firstObject.windowScene.interfaceOrientation == UIInterfaceOrientationMaskLandscape) {
        self.topGradientViewHeightConstraint.constant = 56.0f;
    } else {
        self.topGradientViewHeightConstraint.constant = 100.0f;
    }
    
    self.topGradientViewHeightConstraint.active = YES;

    [self.containerToolBarView setupGradientWithFirstColor:[UIColor.blackColor colorWithAlphaComponent:0.0f] andSecondColor:[UIColor.blackColor colorWithAlphaComponent:0.7f]];
    [self.topGradientView setupGradientWithFirstColor:[UIColor.blackColor colorWithAlphaComponent:0.7f] andSecondColor:[UIColor.blackColor colorWithAlphaComponent:0.0f]];
    
    // configure it if necessary.
    [self configureToolBar];
    
    // configure it if necessary.
    [self configureNavigationBarItems];
}

- (void)setupNavigationBarWillAppear:(BOOL)isWillAppear {
    self.navigationController.navigationBar.barStyle = UIBarStyleBlack;
    self.navigationController.navigationBar.tintColor = UIColor.whiteColor;
    self.navigationController.navigationBar.titleTextAttributes = @{NSForegroundColorAttributeName: UIColor.whiteColor};
    [self.navigationController.navigationBar setBackgroundImage:UIImage.new forBarMetrics:UIBarMetricsDefault];
    self.navigationController.navigationBar.shadowImage = UIImage.new;
    self.navigationController.navigationBar.translucent = isWillAppear;
}

- (void)setupHideToolbarTimerWithTimeInterval:(NSTimeInterval)timeInterval {
    [self invalidateHideToolbarTimer];
    self.toolbarHideTimer =[NSTimer scheduledTimerWithTimeInterval:timeInterval
                                                            target:self
                                                          selector:@selector(hideControls)
                                                          userInfo:nil
                                                           repeats:NO];
    
}

- (void)invalidateHideToolbarTimer {
    if (self.toolbarHideTimer != nil) {
        [self.toolbarHideTimer invalidate];
        self.toolbarHideTimer = nil;
    }
}

- (void)hideControls {
    [self setupControls:NO];
}

- (void)showControls:(BOOL)isShow {
    [self setupControls:isShow];
}

- (void)setupControls:(BOOL)isShow {
    UIColor *color = isShow ? UIColor.whiteColor : UIColor.clearColor;
    
    if (isShow && self.containerToolBarView.hidden) {
        [self setupHideToolbarTimerWithTimeInterval:kHideTimeInterval];
    }
    self.navigationItem.titleView.hidden = !isShow;
    self.navigationItem.leftBarButtonItem.tintColor = color;
    self.navigationItem.rightBarButtonItem.tintColor = color;
    self.navigationController.navigationBar.titleTextAttributes = @{NSForegroundColorAttributeName: color};
    self.topGradientView.hidden = !isShow;
    self.containerToolBarView.hidden = YES;
    self.containerToolBarTopConstraint.constant = isShow == YES ? -126.0f : 0.0f;
    self.containerToolBarView.hidden = !isShow;
    
}

#pragma mark - Transition to size
- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator {
    [super viewWillTransitionToSize:size withTransitionCoordinator:coordinator];
    
    [coordinator animateAlongsideTransition:^(id<UIViewControllerTransitionCoordinatorContext>  _Nonnull context) {
        [self.topGradientView layoutSubviews];
        [self.toolbar layoutSubviews];
        [self reloadContent];
        
        if (UIApplication.sharedApplication.windows.firstObject.windowScene.interfaceOrientation != UIInterfaceOrientationPortrait) {
            self.topGradientViewHeightConstraint.constant = 56.0f;
        } else {
            self.topGradientViewHeightConstraint.constant = 100.0f;
        }
        
    } completion:nil];
}

#pragma mark - UICollectionViewDataSource, UICollectionViewDelegate
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return 0;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    return UICollectionViewCell.new;
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Calls/BaseViewController/BaseViewController.swift
//
//  BaseViewController.swift
//  sample-conference-videochat-swift
//
//  Created by Injoit on 17.08.2020.
//  Copyright © 2020 QuickBlox. All rights reserved.
//

import UIKit

struct BaseConstant {
    static let hideInterval: TimeInterval = 5.0
}

class BaseViewController: UIViewController {
    var collectionView = UICollectionView(frame: .zero, collectionViewLayout: UICollectionViewFlowLayout())
    
    lazy private var containerToolBarView: CallGradientView = {
        let containerToolBarView = CallGradientView(frame: .zero)
        containerToolBarView.backgroundColor = .clear
        return containerToolBarView
    }()
    
    lazy var toolbar: ToolBar = {
        let toolbar = ToolBar(frame: .zero)
        toolbar.backgroundColor = .clear
        toolbar.isTranslucent = true
        return toolbar
    }()
    
    lazy private var topGradientView: CallGradientView = {
        let topGradientView = CallGradientView(frame: .zero)
        topGradientView.backgroundColor = .clear
        return topGradientView
    }()
    
    //MARK: - Internal Properties
    private var toolbarHideTimer: Timer?
    var collectionViewTopConstraint: NSLayoutConstraint!
    private var topGradientViewHeightConstraint: NSLayoutConstraint!
    private var containerToolBarTopConstraint: NSLayoutConstraint!
    
    
    //MARK: - Life Cycle
    init() {
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    deinit {
        debugPrint("deinit \(self)")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        setupNavigationBarWillAppear(true)
        showControls(true)
        setupHideToolbarTimerWithTimeInterval(BaseConstant.hideInterval)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        reloadContent()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        setupNavigationBarWillAppear(false)
        invalidateHideToolbarTimer()
    }
    
    //MARK: - These methods can be overridden in child controllers
    func configureNavigationBarItems() {
        // configure it if necessary.
    }
    
    func configureToolBar() {
        // configure it if necessary.
    }
    
    
    //MARK - Setup Views
    func setupNavigationBarWillAppear(_ isWillAppear: Bool) {
        navigationController?.navigationBar.barStyle = .black
        navigationController?.navigationBar.tintColor = .white
        navigationController?.navigationBar.titleTextAttributes = [NSAttributedString.Key.foregroundColor: UIColor.white]
        navigationController?.navigationBar.setBackgroundImage(UIImage(), for: .default)
        navigationController?.navigationBar.shadowImage = UIImage()
        navigationController?.navigationBar.isTranslucent = isWillAppear
    }
    
    func setupCollectionView() {
       
    }
    
     func configureGUI() {
        view.backgroundColor = .black
        
        // configure it if necessary.
        setupCollectionView()
        
        collectionView.dataSource = self
        collectionView.delegate = self
        view.addSubview(collectionView)
        collectionView.translatesAutoresizingMaskIntoConstraints = false
        let topBarHeight = self.navigationController?.navigationBar.frame.height ?? 44.0
        collectionView.contentInset = UIEdgeInsets(top: -topBarHeight, left: 0, bottom: 0, right: 0)
        collectionViewTopConstraint = collectionView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor)
        collectionViewTopConstraint.constant = -topBarHeight
        collectionViewTopConstraint.isActive = true
        collectionView.leftAnchor.constraint(equalTo: view.leftAnchor).isActive = true
        collectionView.rightAnchor.constraint(equalTo: view.rightAnchor).isActive = true
        collectionView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor).isActive = true
        
        view.addSubview(containerToolBarView)
        containerToolBarView.translatesAutoresizingMaskIntoConstraints = false
        containerToolBarView.leftAnchor.constraint(equalTo: view.leftAnchor).isActive = true
        containerToolBarView.rightAnchor.constraint(equalTo: view.rightAnchor).isActive = true
        containerToolBarTopConstraint = containerToolBarView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -126.0)
        containerToolBarTopConstraint.isActive = true
        containerToolBarView.heightAnchor.constraint(equalToConstant: 126.0).isActive = true
        
        containerToolBarView.addSubview(toolbar)
        toolbar.translatesAutoresizingMaskIntoConstraints = false
        toolbar.leftAnchor.constraint(equalTo: containerToolBarView.leftAnchor).isActive = true
        toolbar.topAnchor.constraint(equalTo: containerToolBarView.topAnchor).isActive = true
        toolbar.rightAnchor.constraint(equalTo: containerToolBarView.rightAnchor).isActive = true
        toolbar.heightAnchor.constraint(equalToConstant: 96.0).isActive = true
        
        view.addSubview(topGradientView)
        topGradientView.translatesAutoresizingMaskIntoConstraints = false
        topGradientView.leftAnchor.constraint(equalTo: view.leftAnchor).isActive = true
        topGradientView.rightAnchor.constraint(equalTo: view.rightAnchor).isActive = true
        topGradientView.topAnchor.constraint(equalTo: collectionView.topAnchor).isActive = true
        topGradientViewHeightConstraint = topGradientView.heightAnchor.constraint(equalToConstant: 100.0)
        if let isLandscape = UIApplication.shared.windows.first?.windowScene?.interfaceOrientation.isLandscape {
            print("isLandscape \(isLandscape)")
            topGradientViewHeightConstraint.constant = 56.0
        } else {
            topGradientViewHeightConstraint.constant = 100.0
        }
        
        topGradientViewHeightConstraint.isActive = true
        
        containerToolBarView.setupGradient(firstColor: UIColor.black.withAlphaComponent(0.0), secondColor: UIColor.black.withAlphaComponent(0.7))
        topGradientView.setupGradient(firstColor: UIColor.black.withAlphaComponent(0.7), secondColor: UIColor.black.withAlphaComponent(0.0))
        
        // configure it if necessary.
        configureToolBar()
        
        // configure it if necessary.
        configureNavigationBarItems()
    }
    
    //MARK: - Public Methods
    func reloadContent() {
        collectionView.reloadData()
    }
    
    //MARK: - Hide/Show Controls Methods
    func setupHideToolbarTimerWithTimeInterval(_ timeInterval: TimeInterval) {
        invalidateHideToolbarTimer()
        self.toolbarHideTimer = Timer.scheduledTimer(timeInterval: timeInterval,
                                                     target: self,
                                                     selector: #selector(hideControls),
                                                     userInfo: nil,
                                                     repeats: false)
    }
    
    @objc private func hideControls() {
        showControls(false)
    }
    
    func showControls(_ isShow: Bool) {
        setupControls(isShow)
    }
    
    func setupControls(_ isShow: Bool) {
        let color: UIColor = isShow == true ? .white : .clear
        if isShow == true, containerToolBarView.isHidden == true {
            setupHideToolbarTimerWithTimeInterval(BaseConstant.hideInterval)
        }
        navigationItem.titleView?.isHidden = !isShow
        navigationItem.leftBarButtonItem?.tintColor = color
        navigationItem.rightBarButtonItem?.tintColor = color
        navigationController?.navigationBar.titleTextAttributes = [NSAttributedString.Key.foregroundColor: color]
        topGradientView.isHidden = !isShow
        containerToolBarView.isHidden = !isShow
        containerToolBarTopConstraint.constant = isShow == true ? -126.0 : 0.0
    }
    
    //MARK: - Internal Methods
    func invalidateHideToolbarTimer() {
        if self.toolbarHideTimer != nil {
            self.toolbarHideTimer?.invalidate()
            self.toolbarHideTimer = nil
        }
    }
    
    // MARK: Transition to size
    override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
        super.viewWillTransition(to: size, with: coordinator)
        
        coordinator.animate(alongsideTransition: { [weak self] context in
            guard let self = self else {return}
            self.topGradientView.layoutSubviews()
            self.toolbar.layoutSubviews()
            self.reloadContent()
            
            if let isLandscape = UIApplication.shared.windows.first?.windowScene?.interfaceOrientation.isLandscape {
                print("isLandscape \(isLandscape)")
                self.topGradientViewHeightConstraint.constant = 56.0
            } else {
                self.topGradientViewHeightConstraint.constant = 100.0
            }
        })
    }
}

extension BaseViewController: UICollectionViewDataSource, UICollectionViewDelegate {
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return 0
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        return UICollectionViewCell()
    }
}
